keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -91,242 +92,324 @@ with the unreasonable development overhead of a webhook.
-              type: integer
-    ```
-    and similarly for `x-kubernetes-deletion-only: true`.
-    
-For list type `atomic` we disallow `x-kubernetes-immutable-keys`.
+```
+       
+Allowed:
+- `{}` → `{foo:[]}` (all keys are unchanged)
+- `{foo:[]}` → `{}` (all keys are unchanged)
+- `{foo:[{k:a,v:1}]}` → `{foo:[{k:a,v:2}]}` (all keys are unchanged)
+- `{foo:[{k:a},{k:b}]}` → `{foo:[{k:b},{k:a}]}` (all keys are unchanged)
+
+Allowed for AddOnly, disallowed otherwise:
+- `{}` → `{foo:[{k:a}]}` (key a is added)
+- `{foo:[{k:a}]}` → `{foo:[{k:a},{k:b}]}` (key b is added)
+
+Allowed for RemoveOnly, disallowed otherwise:
+- `{foo:[{k:a}]}` → `{}` (key a is removed)
+- `{foo:[{k:a}]}` → `{foo:[]}` (key a is removed)
+
+Disallow:
+- `{foo:[{k:a}]}` → `{foo:[{k:b}]}` (key a is removed, key b is added)
+
+#### Immutable array keys, mutable value, list type set
+
+The whole items are the keys. Hence, new items cannot be added (Immutable, RemoveOnly), old items cannot be removed (Immutable, AddOnly), but the order can.
+
+Note: this is different from a set with `x-kubernetes-mutability: Immutable`. The latter does not allow order changes.
+
+*Example 9*: 
 
-Note: it is planned to add `x-kubernetes-map-type` as the equivalent vendor extensions for map to distinguish between classical and atomic maps with the values `map` and `atomic`. The `map` case is the same as the one without any specified map type, and for atomic maps we disallow `x-kubernetes-immutable-keys` again.
+```yaml
+properties:
+  foo:
+    x-kubernetes-list-type: set
+    x-kubernetes-key-mutability: Immutable | AddOnly | RemoveOnly
+    items:
+      type: string
+```
+
+Allowed:
+- `{}` → `{foo: []}` (all keys are unchanged)
+- `{foo: []}` → `{}` (all keys are unchanged)
+- `{foo: [a,b]}` → `{foo: [b,a]}` (all keys are unchanged)
+
+Allowed for AddOnly, disallowed otherwise:
+- `{foo:[a]}` → `{foo:[a,b]}` (key a is unchanged, key b is added)
+- `{foo:[a]}` → `{foo:[b,a]}` (key a is unchanged, key b is added)
+
+Disallowed for RemoveOnly, disallowed otherwise:
+- `{foo:[a]}` → `{foo:[]}` (key a is removed)
+
+Disallowed:
+- `{foo:[a]}` → `{foo:[b]}` (key a is removed, key b is added)
+
+#### Immutable map keys, mutable values, map type map / atomic
+
+Equivalently to the list type map, the set of keys is immutable|add-only|remove-only. Hence, values can be changed. New key-values cannot be added (Immutable, RemoveOnly) and old key-values cannot be removed (Immutable, AddOnly).
+
+*Example 10*:
+
+```yaml
+properties:
+  foo:
+    x-kubernetes-map-type: map | atomic 
+    x-kubernetes-key-mutability: Immutable | AddOnly | RemoveOnly
+    additionalProperties:
+      type: string
+```
+
+Allowed:
+
+- `{}` → `{foo:{}}`	(keys are unchanged)
+- `{foo:{}}` → `{}`	(keys are unchanged)
+- `{foo:{a:"1"}` → `{foo:{a:"2"}}` (keys are unchanged)
+
+Allowed for AddOnly, disallowed otherwise:
+- `{}` → `{foo:{a:"1"}}` (key a is added)
+- `{foo:{a:"1"}}` → `{foo:{a:"1", b:"2"}}` (key b is added)
```

Comment:
this seems incompatible with a `x-kubernetes-map-type: atomic`

Reply:
why?